Add recursive join docs. (#443)
returntocorp/semgrep-docs: docs/experiments/join-mode/recursive-joins.md

# Recursive joins

> Join mode is an extension on top of Semgrep that runs multiple rules at once and only returns results if certain conditions are met. Join mode is an experimental mode that lets you cross file boundaries, allowing you to write rules for whole code bases instead of individual files. More information is available in the docs, here: https://semgrep.dev/docs/experiments/join-mode/

Reviewer: ```suggestion
Join mode is an extension of Semgrep that runs multiple rules at once and only returns results if certain conditions are met. Join mode is an experimental mode that lets you cross file boundaries, allowing you to write rules for whole code bases instead of individual files. More information is available in [these docs](/docs/experiments/join-mode/).
```

Consider the following Python script and rule.

```python
def function_1():
    print("hello")
    function_2()

def function_2():
    function_4()

def function_3():
    function_5()

def function_4():
    function_5()

def function_5():
    print("goodbye")
```

```yaml
rules:
- id: python-callgraph
  message: python callgraph
  languages: [python]
  severity: INFO
  pattern: |
    def $CALLER(...):
      ...
      $CALLEE(...)
```

A join condition like this, `python-callgraph.$CALLER --> python-callgraph.$CALLEE`, will produce a table like this. Notice how `function_1` appears with `function_4` and `function_5` as callees, even though it isn't directly called.

| $CALLER  | $CALLEE  |
|----------|----------|
|function_1|function_2|
|function_1|function_4|
|function_1|function_5|
|function_1|print     |
|function_2|function_4|
|function_2|function_5|
|function_3|function_5|
|function_4|function_5|
|function_5|print     |

## Example Rule

> To really utilize join mode, it's important to think of a join mode rule as "asking questions about the whole project", rather than looking for a single pattern. For example, to find a SQL injection, you need to understand a few things about the project:

Reviewer: ```suggestion
It's important to think of a join mode rule as "asking questions about the whole project", rather than looking for a single pattern. For example, to find a SQL injection, you need to understand a few things about the project:
```

1. Is there any user input?
1. Do any functions manually build a SQL string using function input?
1. Can the user input reach the function that manually builds the SQL string?

> Now, you can write individual Semgrep rules that gather information about each of these questions. This example will use https://github.com/ScaleSec/vulnado as an example for finding a SQL injection. Vulnado is a Spring application.

Reviewer: ```suggestion
Now, you can write individual Semgrep rules that gather information about each of these questions. This example  uses [Vulnado](https://github.com/ScaleSec/vulnado) for finding a SQL injection. Vulnado is a Spring application.
```

> The first rule will look for user input into the Spring application. This rule also captures sinks that use a user-inputtable parameter as an argument.

Reviewer: ```suggestion
The first rule looks for user input into the Spring application. This rule also captures sinks that use a user-inputtable parameter as an argument.
```

        regex: (?i)(select|delete|insert).*
```

Finally, the third rule is used to construct a pseudo-callgraph:

```yaml
rules:
- id: java-callgraph 
  languages: [java]
  severity: INFO
  message: $CALLER calls $OBJ.$CALLEE
  patterns:
  - pattern-inside: |
      $TYPE $CALLER(...) {
        ...
      }
  - pattern: $OBJ.$CALLEE(...)
```

The join rule, at last, looks like this:

```yaml
rules:
- id: spring-sql-injection
  message: SQLi
  severity: ERROR
  mode: join
  join:
    refs:
    - rule: rule_parts/java-spring-user-input.yaml
      as: user-input
    - rule: rule_parts/method-parameter-formatted-sql.yaml
      as: formatted-sql
    - rule: rule_parts/java-callgraph.yaml
      as: callgraph
    on:
    - 'callgraph.$CALLER --> callgraph.$CALLEE'
    - 'user-input.$SINK == callgraph.$CALLER'
    - 'callgraph.$CALLEE == formatted-sql.$METHODNAME'
```

The `on:` conditions, in order, read like this:
- Recursively generate a pseudo callgraph on $CALLER to $CALLEE.
- Match when a method with user input has a $SINK that is the $CALLER in the pseudo-callgraph, and
- Match when the $CALLEE is the $METHODNAME of a method that uses a parameter to construct a SQL string.

> Running this on vulnado will produce tables that look like this:

Reviewer: ```suggestion
Running this on Vulnado will produce tables that look like this:
```
